core/src/main/java/org/jdbi/v3/core/statement/Slf4JSqlLogger.java

```diff
@@ -0,0 +1,80 @@
+package org.jdbi.v3.core.statement;
+
+import java.sql.SQLException;
+import java.time.Duration;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple {@link SqlLogger} that emits some diagnostic information about
+ * Jdbi usage. Usually you shouldn't log every database statement made but
+ * it can be helpful for debugging.
+ */
+public class Slf4JSqlLogger implements SqlLogger {
+    private final Logger log;
+
+    public Slf4JSqlLogger() {
+        this(LoggerFactory.getLogger("org.jdbi.sql"));
+    }
+
+    public Slf4JSqlLogger(Logger log) {
+        this.log = log;
+    }
+
+    @Override
+    public void logAfterExecution(StatementContext context) {
+        if (log.isDebugEnabled()) {
+            log.debug("Executed in {} '{}' with parameters '{}'",
+                    format(Duration.between(context.getExecutionMoment(), context.getCompletionMoment())),
+                    context.getParsedSql().getSql(),
+                    context.getBinding());
+        }
+    }
+
+    @Override
+    public void logException(StatementContext context, SQLException ex) {
+        if (log.isErrorEnabled()) {
+            log.error("Exception while executing '{}' with parameters '{}'",
+                    context.getParsedSql().getSql(),
+                    context.getBinding(),
+                    ex);
+        }
+    }
+
+    private String format(Duration duration) {
+        final long totalSeconds = duration.getSeconds();
+        final StringBuilder result = new StringBuilder();
+        final long h = totalSeconds / 3600;
+        if (h > 0) {
+            result.append(h);
+            result.append(':');
+        }
+
+        final long m = (totalSeconds % 3600) / 60;
+        if (m > 0) {
+            if (m < 10) {
+                result.append('0');
+            }
+            result.append(m);
+            result.append(':');
+        }
+
+        final long s = totalSeconds % 60;
+        if (s < 10) {
+            result.append('0');
+        }
+        result.append(s);
+
+        result.append('.');
+        final long millis = duration.toMillis() % 1000;
+        if (millis < 100) {
+            result.append('0');
+        }
+        if (millis < 10) {
+            result.append('0');
+        }
+        result.append(millis);
+        return result.toString();
```

Comment:
I think a `String.format` here will be simpler. Something like this: `return String.format("%d:%02d:%02d.%03d", h, m, s, millis);` — I'm sure that's not 100% accurate, but you get the idea.

Reply:
Yeah, it is much much slower to call format (20x in a benchmark for this) but I suppose the performance here is not too critical.